runtime/sema/check_variable_declaration.go

```diff
@@ -234,3 +236,119 @@ func (checker *Checker) elaborateNestedResourceMoveExpression(expression ast.Exp
 		checker.Elaboration.IsNestedResourceMoveExpression[expression] = struct{}{}
 	}
 }
+
+func (checker *Checker) recordReferenceCreation(target, expr ast.Expression) {
+	switch target := target.(type) {
+	case *ast.IdentifierExpression:
+		checker.recordReference(target.Identifier.Identifier, expr)
+	default:
+		// Currently it's not possible to track the references
+		// assigned to member-expressions/index-expressions.
+		return
+	}
+}
+
+func (checker *Checker) recordReference(targetVarName string, expr ast.Expression) {
+	referencedVar := checker.referencedVariable(expr)
+
+	if referencedVar != nil &&
+		referencedVar.Type.IsResourceType() {
+		checker.references.Set(targetVarName, referencedVar)
+	}
+}
+
+// referencedVariable return the referenced variable
+func (checker *Checker) referencedVariable(expr ast.Expression) *Variable {
+	refExpression := checker.referenceExpression(expr)
+
+	var variableRefExpr *ast.Identifier
+
+	switch refExpression := refExpression.(type) {
+	case *ast.ReferenceExpression:
+		// If it is a reference expression, then find the root variable.
+		// i.e: root variable of `&a.b.c as &T` is `a`.
+		// This is because nested resources cannot be moved.
+		// If `c` needs to be moved, then `a` has to be moved.
+		// So tracking `a` is sufficient.
+		variableRefExpr = checker.variableReferenceExpression(refExpression.Expression)
+	case *ast.IdentifierExpression:
+		variableRefExpr = &refExpression.Identifier
+	default:
+		return nil
+	}
+
+	if variableRefExpr == nil {
+		return nil
+	}
+
+	referencedVariableName := variableRefExpr.Identifier
+
+	for {
+		// If the referenced variable is again a reference,
+		// then find the variable of the root of the reference chain.
+		// e.g::
+		//     ref1 = &v
+		//     ref2 = &ref1[0]
```

Comment:
Is the indexing an accident?

```suggestion
		//     ref2 = &ref1
```

Reply:
no, what I actually meant there was taking a reference to an inner element, via a reference.
But you are right, we can simplify that by simply taking a reference to a reference.